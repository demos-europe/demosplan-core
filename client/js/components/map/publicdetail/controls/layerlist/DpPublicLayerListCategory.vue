<license>
  (c) 2010-present DEMOS plan GmbH.

  This file is part of the package demosplan,
  for more information see the license file.

  All rights reserved
</license>

<template>
  <li
    v-if="layers.length > 0"
    :class="prefixClass('c-map__group-item')"
    :title="group.attributes.name">
    <div
      :class="[
        isVisible ? prefixClass('is-active') : '',
        prefixClass('c-map__layer')
      ]">
      <span :class="prefixClass('c-map__group-item-controls')">
        <button
          :class="prefixClass('btn--blank btn--focus w-3 text-left')"
          :aria-label="group.attributes.name + ' ' + (isVisible ? Translator.trans('maplayer.category.hide') : Translator.trans('maplayer.category.show'))"
          @click="toggle">
          <i
            :class="[isVisible ? prefixClass('fa-eye') : prefixClass('fa-eye-slash'), prefixClass('fa')]"
            aria-hidden="true" />
        </button>
        <button
          v-if="false === appearsAsLayer"
          :class="prefixClass('btn--blank btn--focus w-3 text-left')"
          :aria-label="group.attributes.name + ' ' + (unfolded ? Translator.trans('maplayer.category.close') : Translator.trans('maplayer.category.open'))"
          @click="fold">
          <i
            :class="[unfolded ? prefixClass('fa fa-folder-open') : prefixClass('fa fa-folder')]"
            aria-hidden="true" />
        </button>

      </span>
      <span
        @click="appearsAsLayer ? toggle : fold"
        :class="prefixClass('c-map__group-item-name o-hellip--nowrap')">
        {{ group.attributes.name }}
      </span>
      <dp-contextual-help
        v-if="'' !== contextualHelp"
        class="c-map__layerhelp"
        :text="contextualHelp" />
    </div>
    <dp-public-layer-list
      :layer-groups-alternate-visibility="layerGroupsAlternateVisibility"
      :layers="layers"
      :parent-is-visible="isVisible"
      :unfolded="unfolded"
      :class="[appearsAsLayer ? prefixClass('sr-only') : prefixClass('c-map__group-item-child u-mr-0')]" />
  </li>
</template>

<script>
import { DpContextualHelp, hasOwnProp, prefixClass } from '@demos-europe/demosplan-ui'
import { mapActions, mapGetters, mapMutations } from 'vuex'
import DpPublicLayerList from './DpPublicLayerList'

export default {
  name: 'DpPublicLayerListCategory',
  components: { DpContextualHelp },

  props: {
    group: {
      type: Object,
      required: true
    },

    layerType: {
      type: String,
      required: false,
      default: 'overlay'
    },

    layerGroupsAlternateVisibility: {
      type: Boolean,
      required: false,
      default: false
    }
  },

  emits: [
    'layer:hideOtherCategories',
    'layer:showParent',
    'layer:toggleChildCategories',
    'layer:toggleChildLayer'
  ],

  data () {
    return {
      unfolded: false,
      appearsAsLayer: this.group.attributes.layerWithChildrenHidden,
      contextualHelp: '',
      tooltipExpanded: false
    }
  },

  computed: {
    ...mapGetters('Layers', ['isLayerVisible', 'element', 'elementListForLayerSidebar', 'rootId']),

    isTopLevelCategory () {
      return this.rootId === this.group.attributes.parentId
    },

    isVisible () {
      return this.isLayerVisible(this.group.id)
    },

    layers () {
      return this.elementListForLayerSidebar(this.group.id, this.layerType, true)
    }
  },

  methods: {
    ...mapMutations('Layers', [
      'setLayerState'
    ]),

    ...mapActions('Layers', ['toggleCategoryAndItsChildren']),

    fold () {
      this.unfolded = (this.unfolded === false)
    },

    prefixClass (classList) {
      return prefixClass(classList)
    },

    // Toggle self and children
    toggle () {
      this.toggleCategoryAndItsChildren({ id: this.group.id, isVisible: !this.isVisible })
    },

    isParentOf (elementList, categoryId) {
      let isParent = false

      for (const key in elementList) {
        //  Skip loop if the property is from prototype
        if (hasOwnProp(elementList, key) === false) continue

        const element = elementList[key]

        //  If the currently looped category is the direct parent of the toggled layer, return here
        if (element.type === 'GisLayerCategory' && element.id === categoryId) {
          return true
        }

        //  If the currently looped category is not the parent of toggled layer, check its child categories
        if (element.type === 'GisLayerCategory' && element.id !== categoryId) {
          const elementList = this.elementListForLayerSidebar(element.id, 'overlay', true)

          isParent = this.isParentOf(elementList, categoryId)
        }
      }

      return isParent
    }
  },

  mounted () {
    if (this.group.attributes.isVisible) {
      this.setLayerState({ id: this.group.id, key: 'isVisible', value: true })
    }

    // Handle data for the category that has to appear as Layer and hides his children
    if (this.group.attributes.layerWithChildrenHidden) {
      this.appearsAsLayer = true
      this.isVisible = this.group.attributes.hasDefaultVisibility
      // Get contextualHelp from all children
      this.layers.forEach(el => {
        const contextualHelp = this.element({ id: el.id, type: 'ContextualHelp' })
        if (hasOwnProp(contextualHelp, 'attributes') && contextualHelp.attributes.text !== '') {
          this.contextualHelp += contextualHelp.attributes.text + ' '
        }
      })
    }
  },

  beforeCreate () {
    this.$options.components.dpPublicLayerList = DpPublicLayerList
  }
}
</script>
