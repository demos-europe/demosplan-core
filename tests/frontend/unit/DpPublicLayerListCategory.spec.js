/**
 * (c) 2010-present DEMOS E-Partizipation GmbH.
 *
 * This file is part of the package demosplan,
 * for more information see the license file.
 *
 * All rights reserved
 */

import DpPublicLayerListCategory from '@DpJs/components/map/publicdetail/controls/layerlist/DpPublicLayerListCategory'
import shallowMountWithGlobalMocks from '@DpJs/VueConfigLocal'

const propsData = {
  group: {
    id: 'group-id',
    attributes: {
      parentId: 'xxx-parentID',
    },
    relationships: {
      categories: {
        data: [],
      },
      gisLayers: {
        data: [],
      },
    },
  },
  layerGroupsAlternateVisibility: true,
}

describe('DpPublicLayerListCategory', () => {
  it('should be an object', () => {
    expect(typeof DpPublicLayerListCategory).toBe('object')
  })

  it('should be named DpPublicLayerListCategory', () => {
    expect(DpPublicLayerListCategory.name).toBe('DpPublicLayerListCategory')
  })

  it.skip('should have the correct prop-values', () => {
    const wrapper = shallowMountWithGlobalMocks(DpPublicLayerListCategory, {
      propsData,
      computed: {
        elementListForLayerSidebar: () => [],
        layers: () => [],
      },
    })

    expect(wrapper.props().layerGroupsAlternateVisibility).toBe(true)
    expect(typeof wrapper.props().group).toBe('object')
    expect(wrapper.props().group.id).toBe('group-id')
    expect(wrapper.props().group.attributes.parentId).toBe('xxx-parentID')
  })

  it.skip('should compute the contextualHelp-ID correctly', () => {
    const layerfromStoreMock = jest.fn()
    layerfromStoreMock.mockReturnValue([{
      layers: [],
      unfolded: false,
      layerType: 'overlay',
      layerGroupsAlternateVisibility: true,
    }, {
      layers: [],
      unfolded: false,
      layerType: 'overlay',
      layerGroupsAlternateVisibility: true,
    }])

    const wrapper = shallowMountWithGlobalMocks(DpPublicLayerListCategory, {
      propsData,
      computed: {
        elementListForLayerSidebar: () => layerfromStoreMock,
      },
      stubs: {
        'dp-public-layer-list': true,
      },
    })

    expect(wrapper.vm.elementListForLayerSidebar()).toEqual([{
      layers: [],
      unfolded: false,
      layerType: 'overlay',
      layerGroupsAlternateVisibility: true,
    }, {
      layers: [],
      unfolded: false,
      layerType: 'overlay',
      layerGroupsAlternateVisibility: true,
    }])

    expect(wrapper.vm.contextualHelpId).toMatch('contextualHelpgroup-id')
    expect(wrapper.vm.id).toMatch('layergroupgroupid')
  })

  it.skip('should compute the isTopLevelCategory correct', () => {
    const layerfromStoreMock = jest.fn()
    layerfromStoreMock.mockReturnValue([{
      layers: [],
      unfolded: false,
      layerType: 'overlay',
      layerGroupsAlternateVisibility: true,
    }, {
      layers: [],
      unfolded: false,
      layerType: 'overlay',
      layerGroupsAlternateVisibility: true,
    }])

    const wrapper = shallowMountWithGlobalMocks(DpPublicLayerListCategory, {
      propsData,
      computed: {
        elementListForLayerSidebar: () => jest.fn().mockReturnValue([]),
        rootId: () => 'xxx-rootID',
      },
      stubs: {
        'dp-public-layer-list': true,
      },
    })

    expect(wrapper.vm.isTopLevelCategory).toBe(false)
  })

  it.skip('should render an empty layout if there are no layers', () => {
    const wrapper = shallowMountWithGlobalMocks(DpPublicLayerListCategory, {
      propsData,
      computed: {
        elementListForLayerSidebar: () => [],
        layers: () => [],
      },
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it.skip('should render a list if there are layers', () => {
    const wrapper = shallowMountWithGlobalMocks(DpPublicLayerListCategory, {
      propsData,
      computed: {
        elementListForLayerSidebar: () => [],
        layers: () => [{
          layers: [],
          unfolded: false,
          layerType: 'overlay',
          layerGroupsAlternateVisibility: true,
        }, {
          layers: [],
          unfolded: false,
          layerType: 'overlay',
          layerGroupsAlternateVisibility: true,
        }],
      },
      stubs: {
        'dp-public-layer-list': true,
      },
    })

    expect(wrapper.html()).toMatchSnapshot()
  })

  it.skip('should toggle its visibility when toggle is called', () => {
    const wrapper = shallowMountWithGlobalMocks(DpPublicLayerListCategory, {
      propsData,
      computed: {
        elementListForLayerSidebar: () => [],
        layers: () => [],
      },
      stubs: {
        'dp-public-layer-list': true,
      },
    })

    wrapper.setData({
      isVisible: true,
    })

    expect(wrapper.vm.isVisible).toBe(true)

    wrapper.vm.toggle()
    expect(wrapper.vm.isVisible).toBe(false)
  })

  it.skip('should toggle its visibility when toggleFromSelf is called', () => {
    const wrapper = shallowMountWithGlobalMocks(DpPublicLayerListCategory, {
      propsData,
      computed: {
        elementListForLayerSidebar: () => [],
        layers: () => [],
      },
      stubs: {
        'dp-public-layer-list': true,
      },
    })

    wrapper.setData({
      isVisible: true,
    })

    expect(wrapper.vm.isVisible).toBe(true)

    wrapper.vm.toggleFromSelf()
    expect(wrapper.vm.isVisible).toBe(false)
  })

  it.skip('should toggle its visibility when toggleFromParent is called', () => {
    const wrapper = shallowMountWithGlobalMocks(DpPublicLayerListCategory, {
      propsData,
      computed: {
        elementListForLayerSidebar: () => [],
        layers: () => [],
      },
      stubs: {
        'dp-public-layer-list': true,
      },
    })

    wrapper.vm.toggleFromParent([], false)
    expect(wrapper.vm.isVisible).toBe(true)
  })

  it.skip('should be parent of a child', () => {
    const elementList = [
      { id: 'aaa', type: 'GisLayer' },
      { id: 'bbb', type: 'GisLayerCategory' },
      { id: 'ccc', type: 'GisLayerCategory' },
      { id: 'ddd', type: 'GisLayerCategory' },
      { id: 'eee', type: 'GisLayer' },
    ]

    const wrapper = shallowMountWithGlobalMocks(DpPublicLayerListCategory, {
      propsData,
      computed: {
        elementListForLayerSidebar: () => { return () => [] },
        layers: () => [],
      },
      stubs: {
        'dp-public-layer-list': true,
      },
    })

    const component = wrapper.vm
    expect(typeof component.elementListForLayerSidebar).toBe('function')
    expect(component.isParentOf(elementList, 'ccc')).toBe(true)
    expect(component.isParentOf(elementList, 'eee')).toBe(false)
    expect(component.isParentOf(elementList, 'ggg')).toBe(false)
  })
})
