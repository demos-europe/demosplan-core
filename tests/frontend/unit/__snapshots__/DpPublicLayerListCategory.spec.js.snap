// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DpPublicLayerListCategory should render a list if there are layers 1`] = `
"<li class=\\"dp-c-map__group-item\\">
  <div class=\\"dp-is-active dp-c-map__layer\\"><span class=\\"dp-c-map__group-item-controls\\"><button aria-label=\\"undefined maplayer.category.hide\\" class=\\"dp-btn--blank dp-btn--focus dp-width-16 dp-text-left\\"><i aria-hidden=\\"true\\" class=\\"dp-fa-eye dp-fa\\"></i></button> <!----></span> <span class=\\"dp-c-map__group-item-name dp-o-hellip--nowrap\\">

    </span>
    <!---->
  </div>
  <dp-public-layer-list-stub layer-groups-alternate-visibility=\\"true\\" layers=\\"[object Object],[object Object]\\" class=\\"dp-c-map__group-item-child dp-u-mr-0\\"></dp-public-layer-list-stub>
</li>"
`;

exports[`DpPublicLayerListCategory should render an empty layout if there are no layers 1`] = `""`;
